crates/rpc/rpc/src/eth/api/mod.rs

```diff
@@ -438,4 +453,144 @@ struct EthApiInner<Provider, Pool, Network> {
     pending_block: Mutex<Option<PendingBlock>>,
     /// A pool dedicated to blocking tasks.
     blocking_task_pool: BlockingTaskPool,
+    /// Cache for block fees history
+    fee_history_cache: FeeHistoryCache<Provider>,
+}
+
+/// Settings for the [EthStateCache](crate::eth::cache::EthStateCache).
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FeeHistoryCacheConfig {
+    /// Max number of blocks in cache.
+    ///
+    /// Default is 1024.
+    pub max_blocks: u64,
+}
+
+impl Default for FeeHistoryCacheConfig {
+    fn default() -> Self {
+        FeeHistoryCacheConfig { max_blocks: 1024 }
+    }
+}
+
+/// Wrapper struct for BTreeMap
+#[derive(Debug, Clone)]
+pub struct FeeHistoryCache<Provider> {
+    lower_bound: Arc<AtomicU64>,
+    upper_bound: Arc<AtomicU64>,
+    config: FeeHistoryCacheConfig,
+    provider: Provider,
+    entries: Arc<tokio::sync::RwLock<BTreeMap<u64, SealedHeader>>>,
+}
+
+impl<Provider> FeeHistoryCache<Provider>
+where
+    Provider: BlockReaderIdExt
+        + ChainSpecProvider
+        + StateProviderFactory
+        + EvmEnvProvider
+        + BlockNumReader
+        + 'static,
+{
+    /// Creates new FeeHistoryCache instance, initialize it with the mose recent data, set bounds
+    pub fn new(config: FeeHistoryCacheConfig, provider: Provider) -> Self {
+        let mut init_tree_map = BTreeMap::new();
+
+        let last_block_number = provider.last_block_number().unwrap_or(0);
+
+        let start_block = if last_block_number > config.max_blocks {
+            last_block_number - config.max_blocks
+        } else {
+            0
+        };
+
+        let headers =
+            provider.sealed_headers_range(start_block..=last_block_number).unwrap_or_default();
+
+        let (mut lower_bound, mut upper_bound) = (0, 0);
+        for header in headers {
+            init_tree_map.insert(header.number, header.clone());
+            upper_bound = header.number;
+            if lower_bound == 0 {
+                lower_bound = header.number;
+            }
+        }
+
+        let entries = Arc::new(tokio::sync::RwLock::new(init_tree_map));
+
+        let upper_bound = Arc::new(AtomicU64::new(upper_bound));
+        let lower_bound = Arc::new(AtomicU64::new(lower_bound));
+
+        FeeHistoryCache { config, entries, provider, upper_bound, lower_bound }
+    }
+
+    /// Processing of the arriving blocks
+    pub async fn on_new_block(&self, block: &SealedBlock) {
```

Comment:
this should take an iterator of blocks and convert the insert process into a loop, so we don't need to acquire a lock multiple times when inserting new blocks